waiter/src/waiter/process_request.clj

```diff
@@ -252,13 +252,26 @@
                                          (let [complete-trigger-id (utils/unique-identifier)]
                                            (deliver complete-triggered-promise complete-trigger-id)
                                            (when (= complete-trigger-id @complete-triggered-promise)
-                                             (log/debug "closing request body as" message)
+                                             ;; avoid decrementing the counter multiple times
                                              (counters/dec! request-body-streaming-counter)
-                                             (when throwable
-                                               (log/error throwable "unable to stream request bytes, aborting request")
-                                               (async/>!! abort-ch throwable))
-                                             (report-request-size-metrics 0 true)
-                                             (async/close! body-ch))))))]
+                                             (report-request-size-metrics 0 true))
+                                           (if throwable
+                                             (let [identifier-map {:identifier complete-trigger-id}
+                                                   ;; the callback is necessary as there is a data race between aborting the
+                                                   ;; request and closing of the body channel triggering a normal complete before
+                                                   ;; the abort request gets processed.
+                                                   callback (fn complete-request-streaming-abort-result-callback [aborted?]
+                                                              (cid/with-correlation-id
+                                                                correlation-id
+                                                                (log/info "result of aborted request:" aborted? identifier-map)
+                                                                (async/close! body-ch)))]
+                                               (log/info throwable "aborting request" message identifier-map)
+                                               (when-not (async/put! abort-ch [throwable callback])
```

Comment:
Why do we need to close the `body-ch` here?  The callback includes closing `body-ch`.  Doesn't the callback get called always?

Reply:
If the put fails, the callback will not be invoked, we want to go ahead and close the channel regardless as we are done streaming the request body.